File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
const solution = (begin, target, words) => {
    if (!words.includes(target)) return 0;
    
    const queue = [[begin, 0]];
    
    while(queue.length) {
        let [cur, count] = queue.shift();
        if(cur === target) return count;
        for(let i=0; i<words.length; i++) {
            if(canConvert(words[i], cur)) {
                queue.push([words[i], count+1]);
            }
        }
    }
    return 0;
}

// 변환 가능한지
const canConvert = (a, b) => {
    let cnt = 0;
    for(let i=0; i<a.length; i++) {
        if(a[i] == b[i]) cnt++;
    }
    if(cnt === a.length-1) return true;
    else return false;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

# 단어가 한개만 다른지 판별하는 함수
def isDiffOne(word1, word2):
    n = len(word1)
    diff = 0
    
    for index in range(n):
        if word1[index] != word2[index]:
            diff += 1
        if diff > 1:
            return False
        
    if diff == 1:
        return True
    else:
        return False

        
    
# BFS
def solution(begin, target, words):

    visited = [False for _ in range(len(words))]
    queue = deque([(begin, 0)])
    
    while queue:
        curWord, dist = queue.popleft()
            
        # 만약 꺼낸 단어가 target이면 종료!
        if curWord == target:
            return dist
        

        
        for index, word in enumerate(words):
            # 딱 한글자만 다르고 방문하지 않은 단어들을 모두 방문한다.
            if isDiffOne(curWord, word) and not visited[index]:
                queue.append((word, dist + 1))
                visited[index] = True

    return 0